SSA/Core/Framework/Refinement.lean

```diff
@@ -0,0 +1,159 @@
+end RefinementNotation
+open RefinementNotation
+
+/--
+The homogenous version of `HRefinement`.
+This enables the notation `a ⊑ b`, where `a, b : α`.
+
+NOTE: This typeclass is not intended for dialect implementors. Please implement
+`DialectRefinement` instead.
+-/
+class Refinement (α : Type) where
+  IsRefinedBy : α → α → Prop
+
+instance [Refinement α] : HRefinement α α where
+  IsRefinedBy := Refinement.IsRefinedBy
+
+/-- Equality induces a trivial (homogenous) refinement relation on any type `α`. -/
+def Refinement.ofEq : Refinement α where
+  IsRefinedBy := Eq
+
+/-! ### Dialect Refinement -/
+
+/--
+`DialectHRefinement` defines an heterogenous refinement relation accross two dialects.
+
+Various instances of the ` ⊑ ` refinement notation will be derived from `DialectHRefinement`.
+In particular, this class defines refinement for monadic values, from which
+refinement of pure values `x, y` is defined as `pure x ⊑ pure y.`
+-/
+class DialectHRefinement (d : Dialect) (d' : Dialect) [TyDenote d.Ty] [TyDenote d'.Ty] where
+  /--
+  We say that `a` is refined by `b`, written as `a ⊑ b`, when
+  every observable behaviour of `b` is allowed by `a`.
+  -/
+  IsRefinedBy {t : d.Ty} {u : d'.Ty} : d.m ⟦t⟧ → d'.m ⟦u⟧ → Prop
+open DialectHRefinement
+
+namespace DialectHRefinement
+variable {d d' : Dialect} [TyDenote d.Ty] [TyDenote d'.Ty] [DialectHRefinement d d']
+variable {t : d.Ty} {u : d'.Ty}
+
+/-- Refinement for monadic values -/
+instance instRefinementMonadic : HRefinement (d.m ⟦t⟧) (d'.m ⟦u⟧) where
+  IsRefinedBy := DialectHRefinement.IsRefinedBy
+
+variable [Pure d.m] [Pure d'.m]
+
+/-- Refinement for pure values -/
+instance instRefinementPure : HRefinement ⟦t⟧ ⟦u⟧ where
+  IsRefinedBy x y := (pure x : d.m _) ⊑ (pure y : d'.m _)
+
+/-- Refinement for *potentially* monadic values -/
+instance instRefinementEffect {eff eff' : EffectKind} :
+    HRefinement (eff.toMonad d.m ⟦t⟧) (eff'.toMonad d'.m ⟦u⟧) where
+  IsRefinedBy x y := eff.coe_toMonad x ⊑ eff'.coe_toMonad y
+
+section Lemmas
+
+@[simp] theorem toMonad_pure_IsRefinedBy_toMonad_pure_iff
+    {x : EffectKind.pure.toMonad d.m ⟦t⟧} {y : EffectKind.pure.toMonad d'.m ⟦u⟧} :
+    (x ⊑ y) ↔ (HRefinement.IsRefinedBy (α := ⟦t⟧) (β := ⟦u⟧) x y) := by
+  rfl
+
+@[simp] theorem toMonad_impure_IsRefinedBy_toMonad_impure_iff
+    {x : EffectKind.impure.toMonad d.m ⟦t⟧} {y : EffectKind.impure.toMonad d'.m ⟦u⟧} :
+    (x ⊑ y) ↔ (HRefinement.IsRefinedBy (α := d.m ⟦t⟧) (β := d'.m ⟦u⟧) x y) := by
+  rfl
+
+end Lemmas
+
+end DialectHRefinement
+
+/--
+A lawful homogenous (i.e., within a single dialect) refinement instance is one
+where refinement is reflexive and transitive (i.e., it is a preorder).
+-/
+class LawfulDialectRefinement (d : Dialect) [TyDenote d.Ty] [DialectHRefinement d d] where
+  isRefinedBy_rfl : ∀ {t : d.Ty} (x : d.m ⟦t⟧), x ⊑ x
+  isRefinedBy_trans : ∀ {t u v : d.Ty} (x : d.m ⟦t⟧) (y : d.m ⟦u⟧) (z : d.m ⟦v⟧),
+    x ⊑ y → y ⊑ z → x ⊑ z
```

Comment:
If an honestly "lawful" dialect refinement better be monotone, then perhaps we rename this to `PreorderDialectRefinement`, and the other notion as `MonotoneDialectRefinement` (defined in the appropriate fle), and then define `Lawful = Preorder + Monotone`?


```suggestion
class PreorderDialectRefinement (d : Dialect) [TyDenote d.Ty] [DialectHRefinement d d] where
  isRefinedBy_rfl : ∀ {t : d.Ty} (x : d.m ⟦t⟧), x ⊑ x
  isRefinedBy_trans : ∀ {t u v : d.Ty} (x : d.m ⟦t⟧) (y : d.m ⟦u⟧) (z : d.m ⟦v⟧),
    x ⊑ y → y ⊑ z → x ⊑ z
```